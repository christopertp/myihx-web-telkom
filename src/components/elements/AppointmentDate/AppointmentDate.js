import React, { useState, useEffect } from 'react';
import moment from 'moment';
import PropTypes from 'prop-types';
import Heading from '../Heading';
import styles from './styles.scoped.css';

export default function AppointmentDate(props) {
  const { className, date, isLoading, mini, setPayload } = props;
  const [arrival, setArrival] = useState('0');
  const [selected, setSelected] = useState('0');
  const customClass = [styles.root, className].filter(Boolean).join(' ');

  const setCleanPayload = p => {
    setPayload({
      bookingId: p.bookingId,
      crewId: p.crewId,
      information: p.information,
      timeSlot: p.timeSlot,
    });
  };

  const onChangeArrival = e => {
    setArrival(e.target.value);
    setCleanPayload({ ...date[selected].arrival[e.target.value].data });
  };

  const onChangeDate = e => {
    const arv = date[e.target.value].arrival[0] ? '0' : '1';
    setArrival(arv);
    setSelected(e.target.value);
    setCleanPayload({ ...date[e.target.value].arrival[arv].data });
  };

  useEffect(() => {
    if (date[0]) {
      const arv = `${date[selected].arrival.findIndex(i => i.availability)}`;
      setCleanPayload({ ...date[0].arrival[arv].data });
      setArrival(arv);
    }
  }, [date, selected]);

  const DateItem = (() => {
    if (isLoading) {
      return [...Array.from({ length: 3 }).keys()].map(i => <div className="loading" key={i} />);
    }
    if (!date.length) {
      return ['No date available'];
    }
    return date.map((i, idx) => (
      <label className={selected === `${idx}` ? styles.selected : ''} key={idx}>
        <input name="date" onChange={onChangeDate} type="checkbox" value={`${idx}`} />
        <span>{moment(i.date).format('ddd')}</span>
        <h4>{moment(i.date).format('D')}</h4>
        <span>{moment(i.date).format('MMM')}</span>
      </label>
    ));
  })();

  const ArrivalItem = (() => {
    if (isLoading) {
      return [...Array.from({ length: 2 }).keys()].map(i => <div className="loading" key={i} />);
    }
    if (!date.length) {
      return [];
    }

    const items = date[selected].arrival;
    const names = mini ? ['08:00 - 10:00', '10:00 - 12:00', '13:00-15:00', '15:00-17:00'] : ['Morning', 'Afternoon'];

    return items.map((i, idx) => (
      <label className={arrival === `${idx}` ? styles.selected : ''} disabled={!i || !i.availability} key={idx}>
        {i && <input checked={arrival === `${idx}`} name="arrival" onChange={onChangeArrival} type="radio" value={`${idx}`} />}
        <h4>{names[idx]}</h4>
        {(items[1] && !mini) && <p>Technician will come around {items[1].slot}</p>}
      </label>
    ));
  })();

  return (
    <section className={customClass}>
      <Heading>Available Dates</Heading>
      <div className={styles.date}>{DateItem}</div>
      <Heading>Estimated Arrival</Heading>
      <div className={[styles.arrival, mini && styles.mini].filter(Boolean).join(' ')}>{ArrivalItem}</div>
      <p>
        Time range is the estimated arrival time.
        You will receive a notification when the technician is heading to your location.
      </p>
    </section>
  );
}

AppointmentDate.defaultProps = {
  className: '',
  date: [],
  isLoading: false,
  mini: false,
  setPayload: () => {},
};

AppointmentDate.propTypes = {
  className: PropTypes.string,
  date: PropTypes.arrayOf(PropTypes.shape({
    date: PropTypes.string,
    arrival: PropTypes.array,
  })),
  isLoading: PropTypes.bool,
  mini: PropTypes.bool,
  setPayload: PropTypes.func,
};
