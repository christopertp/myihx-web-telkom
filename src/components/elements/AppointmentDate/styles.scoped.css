.root > h3 {
  margin-bottom: 0.5rem;
}

.arrival {
  display: grid;
  grid-gap: 1rem;
  grid-template-columns: 1fr 1fr;
  margin-bottom: 1rem;
}

.arrival > div, .arrival > label {
  border: 0.0625rem solid #e5e6e6;
  border-radius: 0.5rem;
  flex: 1;
}

.arrival > div {
  height: 7rem;
}

.arrival > label {
  background-color: #ffffff;
  cursor: pointer;
  padding: 2rem 1rem;
  transition: border 0.25s, box-shadow 0.25s;
}

.arrival > label[disabled] {
  background-color: #e5e5e5 !important;
  border: none !important;
  box-shadow: none !important;
  cursor: default !important;
}

.arrival > label.selected, .arrival > label:hover {
  border: 0.0625rem solid #0eb37e;
}

.arrival > label.selected {
  box-shadow: 0 0.5rem 1rem rgba(0, 0, 1, 0.15);
}

.arrival > label > input { display: none; }

.arrival > label > h4 {
  color: #191a1a;
  font-size: 1rem;
  margin-bottom: 0.25rem;
}

.arrival > label > p, .arrival + p {
  color: #66686a;
  font-size: 0.9rem;
}

.arrival.mini > label.selected {
  background-color: #0eb37e;
}

.arrival.mini > label > h4 {
  margin: 0;
  font-weight: 400;
  text-align: center;
}

.arrival.mini > label.selected > h4 {
  color: #ffffff;
  font-weight: 500;
}

.date {
  color: var(--primary-color);
  display: flex;
  margin-bottom: 0.5rem;
  overflow-x: auto;
  padding-bottom: 1rem;
  scroll-snap-type: x mandatory;
}

.date > div, .date > label {
  border-radius: 0.25rem;
  margin-right: 0.5rem;
}

.date > div {
  height: 7rem;
  width: 8.25rem;
}

.date > label {
  background-color: #ffffff;
  box-shadow: 0 0.5rem 1rem rgba(0, 0, 1, 0.15);
  cursor: pointer;
  padding: 0.5rem 3rem;
  scroll-snap-align: start;
  text-align: center;
}

.date > label.selected { background-color: var(--primary-color); }

.date > label.selected > * { color: #ffffff !important; }

.date > label:last-child { margin-right: 0; }

.date > label > input { display: none; }

.date > label > span {
  font-size: 1rem;
  color: #66686a;
}

.date > label > span:last-of-type { text-transform: uppercase; }

.date > label > h4 {
  color: #191a1a;
  font-size: 2rem;
}

@media only screen and (max-width: 48em) {
  .arrival > div { height: 4.5rem; }  
  .arrival > label { padding: 0.75rem 1rem; }
  .arrival > label > h4 { font-size: 0.8rem; }
  .arrival > label > p, .arrival + p { font-size: 0.7rem; }

  .date { font-size: 0.8rem; }
  .date > div {
    height: 4.5rem;
    width: 3.5rem;
  }  
  .date > label { padding: 0.25rem 1rem }
  .date > label > span { font-size: 0.75rem; }
  .date > label > h4 { font-size: 1.5rem; }
}
