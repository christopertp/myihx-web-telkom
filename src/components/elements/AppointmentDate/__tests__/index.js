import React, { useState } from 'react';
import ShallowRenderer from 'react-test-renderer/shallow';
import AppointmentDate from '../AppointmentDate';

const arrival = [
  { availability: true, data: {}, slot: 'slot-1' },
  { availability: true, data: {}, slot: 'slot-2' },
];
const date = [{ arrival }, { arrival }];

describe('src/components/elements/AppointmentDate', () => {
  test('render', () => {
    const shallow = new ShallowRenderer();
    const tree = shallow.render(<AppointmentDate date={date} mini />);
    expect(tree).toMatchSnapshot();
    AppointmentDate.defaultProps.setPayload();
  });

  test('loading state', () => {
    const props = { ...AppointmentDate.defaultProps, isLoading: true };
    const result = AppointmentDate(props);
    expect(result.props.children[1].props.children[0].props.className).toBe('loading');
    expect(result.props.children[3].props.children[0].props.className).toBe('loading');
  });

  test('no data state', () => {
    const result = AppointmentDate(AppointmentDate.defaultProps);
    expect(result.props.children[1].props.children[0]).toBe('No date available');
    expect(result.props.children[3].props.children.length).toBe(0);
  });

  test('onChangeArrival', () => {
    const target = { value: '0' };
    const setArrival = jest.fn();
    useState.mockImplementationOnce(() => ['1', setArrival]);

    const setPayload = jest.fn();

    const props = { ...AppointmentDate.defaultProps, date, setPayload };
    const result = AppointmentDate(props);
    const input = result.props.children[3].props.children[0].props.children[0];
    input.props.onChange({ target });
    expect(setArrival).toHaveBeenCalledWith('0');
    expect(setPayload).toHaveBeenCalled();
  });

  test('onChangeDate', () => {
    const target = { value: '0' };
    const setArrival = jest.fn();
    const setSelected = jest.fn();
    useState
      .mockImplementationOnce(() => ['1', setArrival])
      .mockImplementationOnce(() => ['1', setSelected]);

    const setPayload = jest.fn();
    const arrival = [
      '',
      { availability: true, data: {}, slot: 'slot-2' },
    ];
    const arrival2 = [
      { availability: true, data: {}, slot: 'slot-1' },
      { availability: true, data: {}, slot: 'slot-2' },
    ];
    const date = [{ arrival }, { arrival: arrival2 }];

    const props = { ...AppointmentDate.defaultProps, date, setPayload };
    const result = AppointmentDate(props);
    const input = result.props.children[1].props.children[0].props.children[0];
    input.props.onChange({ target });
    expect(setArrival).toHaveBeenNthCalledWith(2, '1');
    expect(setSelected).toHaveBeenNthCalledWith(1, '0');

    input.props.onChange({ target: { value: '1' } });
    expect(setArrival).toHaveBeenNthCalledWith(3, '0');
    expect(setSelected).toHaveBeenNthCalledWith(2, '1');
    expect(setPayload).toHaveBeenCalledTimes(3);
  });
});
