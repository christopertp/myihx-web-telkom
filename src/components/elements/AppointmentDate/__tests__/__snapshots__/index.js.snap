// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`src/components/elements/AppointmentDate render 1`] = `
<section
  className=""
>
  <Heading
    className=""
  >
    Available Dates
  </Heading>
  <div>
    <label>
      <input
        name="date"
        onChange={[Function]}
        type="checkbox"
        value="0"
      />
      <span>
        ddd
      </span>
      <h4>
        D
      </h4>
      <span>
        MMM
      </span>
    </label>
    <label
      className=""
    >
      <input
        name="date"
        onChange={[Function]}
        type="checkbox"
        value="1"
      />
      <span>
        ddd
      </span>
      <h4>
        D
      </h4>
      <span>
        MMM
      </span>
    </label>
  </div>
  <Heading
    className=""
  >
    Estimated Arrival
  </Heading>
  <div
    className=""
  >
    <label
      disabled={false}
    >
      <input
        checked={true}
        name="arrival"
        onChange={[Function]}
        type="radio"
        value="0"
      />
      <h4>
        08:00 - 10:00
      </h4>
    </label>
    <label
      className=""
      disabled={false}
    >
      <input
        checked={false}
        name="arrival"
        onChange={[Function]}
        type="radio"
        value="1"
      />
      <h4>
        10:00 - 12:00
      </h4>
    </label>
  </div>
  <p>
    Time range is the estimated arrival time. You will receive a notification when the technician is heading to your location.
  </p>
</section>
`;
